### ISSUE TITLE
EmrEtlRunner: add S3DistpCp step for thrift files in processing 

### ISSUE THREAD

**Issue by alexanderdean** [snowplow/snowplow#1647]
See this thread: https://groups.google.com/forum/#!topic/snowplow-user/nLjOuKwZoIQ

We have successfully reproduced this on snplow5 with lzo s3 sink and EMR enrichment. Notes:

* We are running in a subnet, unlike Michael
* Have tried deleting the .lzo.index files and running, and it still fails - suggests the problem isn't to do with reading the .lzo.index files

Next test: run the snplow5 EMR process with some vanilla CloudFront Collector files to see if it's something magical about the .lzo files.

**Comment by epylinkn** [comment]
:+1:

**Comment by mthomas** [comment]
:+1:

**Comment by alexanderdean** [comment]
Actually the error we get is slightly different:

```
Caused by: Status Code: 403, AWS Service: Amazon S3, AWS Request ID: F6003FA50CD80573, AWS Error Code: null, AWS Error Message: Forbidden, S3 Extended Request ID: ...
```

I have tried renaming the files to have shorter names and that didn't work. I'm going to try the CloudFront files next...

**Comment by alexanderdean** [comment]
The difference of error was because we were using an S3 file in enrichments which we didn't have access to. With that removed, we get the exact same error as Michael and Anthony. This is good because we are running in a VPC subnet, and Michael isn't, so that rules out it being somehow VPC-related.

I'm investigating...

**Comment by alexanderdean** [comment]
This is the error I get:

```
java.lang.IllegalArgumentException: AWS Access Key ID and Secret Access Key must be specified as the username or password (respectively) of a s3n URL, or by setting the fs.s3n.awsAccessKeyId or fs.s3n.awsSecretAccessKey properties (respectively).
	at org.apache.hadoop.fs.s3.S3Credentials.initialize(S3Credentials.java:66)
	at org.apache.hadoop.fs.s3native.Jets3tNativeFileSystemStore.initialize(Jets3tNativeFileSystemStore.java:49)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.apache.hadoop.io.retry.RetryInvocationHandler.invokeMethod(RetryInvocationHandler.java:82)
	at org.apache.hadoop.io.retry.RetryInvocationHandler.invoke(RetryInvocationHandler.java:59)
	at org.apache.hadoop.fs.s3native.$Proxy7.initialize(Unknown Source)
	at org.apache.hadoop.fs.s3native.NativeS3FileSystem.initialize(NativeS3FileSystem.java:216)
	at org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:1386)
	at org.apache.hadoop.fs.FileSystem.access$200(FileSystem.java:66)
	at org.apache.hadoop.fs.FileSystem$Cache.get(FileSystem.java:1404)
	at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:254)
	at org.apache.hadoop.fs.Path.getFileSystem(Path.java:187)
	at com.twitter.elephantbird.mapreduce.input.MultiInputFormat.determineFileFormat(MultiInputFormat.java:179)
	at com.twitter.elephantbird.mapreduce.input.MultiInputFormat.createRecordReader(MultiInputFormat.java:88)
	at com.twitter.elephantbird.mapred.input.DeprecatedInputFormatWrapper$RecordReaderWrapper.<init>(DeprecatedInputFormatWrapper.java:251)
	at com.twitter.elephantbird.mapred.input.DeprecatedInputFormatWrapper.getRecordReader(DeprecatedInputFormatWrapper.java:118)
	at cascading.tap.hadoop.io.MultiInputFormat$1.operate(MultiInputFormat.java:253)
	at cascading.tap.hadoop.io.MultiInputFormat$1.operate(MultiInputFormat.java:248)
	at cascading.util.Util.retry(Util.java:762)
	at cascading.tap.hadoop.io.MultiInputFormat.getRecordReader(MultiInputFormat.java:247)
	at org.apache.hadoop.mapred.MapTask$TrackedRecordReader.<init>(MapTask.java:197)
	at org.apache.hadoop.mapred.MapTask.runOldMapper(MapTask.java:418)
	at org.apache.hadoop.mapred.MapTask.run(MapTask.java:372)
	at org.apache.hadoop.mapred.Child$4.run(Child.java:255)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:415)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1121)
	at org.apache.hadoop.mapred.Child.main(Child.java:249)
```

It's the exact same as Anthony's. Michael's is odd because it doesn't look like the error is caused when trying to read the lzo files.

**Comment by alexanderdean** [comment]
Right so I believe the problem is AWS's recent IAM roles policy - see this thread:

https://github.com/awslabs/emr-bootstrap-actions/issues/33

> When you use IAM roles, core-site.xml will not have these things:
> fs.s3n.awsAccessKeyId
>  fs.s3.awsAccessKeyId
> fs.s3bfs.awsAccessKeyId

And it looks like the hadoop-lzo code needs those properties to read the lzo files. I'll update the ticket name.

**Comment by alexanderdean** [comment]
Here is more information on this: http://docs.aws.amazon.com/ElasticMapReduce/latest/DeveloperGuide/emr-iam-roles-calling.html

We can probably fix this by adding an S3DistCp step in front of the lzo files, so that hadoop-lzo doesn't have to attempt S3 access...

**Comment by alexanderdean** [comment]
Right so I have the S3DistCp in place. The lzo filenames were causing some trouble with S3DistCp - apparently the version of S3DistCp on AMI 2.4.2 has:

> some difficulties with copying S3 objects that may have similar paths (prefixes). AMI 2.4.3 and later > has a fixed version of S3DistCP. 

Trying 2.4.3 gave me an error in the Iglu code (something related to Jackson parsing, maybe a Jackson dependency changed between 2.4.2 and 2.4.3). I am trying 2.4.5 now. If that fails, I am going to attempt renaming the lzo files and seeing if that fixes it.

**Comment by alexanderdean** [comment]
So I can't promote to 2.4.3 because of this issue: https://github.com/snowplow/snowplow/issues/1001

I'm going to experiment with a different filename format for the lzo files and see if that works with 2.4.2...

**Comment by alexanderdean** [comment]
So I tried another filename format (UUIDs) and that didn't work either - the problem is that whatever filename format I use, there are always two files with almost identical names (`.lzo`) and (`.lzo.index`)...

So I think the only solution is to fix #1001.

**Comment by danielzohar** [comment]
:+1:
